polar-core/src/sugar.rs

```diff
@@ -0,0 +1,850 @@
+use std::collections::{HashMap, HashSet};
+
+use super::error::{ParseError, PolarResult, RuntimeError};
+use super::kb::KnowledgeBase;
+use super::rules::*;
+use super::terms::*;
+
+// TODO(gj): disallow same string to be declared as a perm/role and a relation.
+// This'll come into play for "owner"-style actor relationships.
+
+#[derive(Clone, Debug)]
+pub enum Declaration {
+    Role,
+    Permission,
+    Relation(Term),
+}
+
+type Declarations = HashMap<Term, Declaration>;
```

Comment:
> I'm assuming you stored Term here because you want to keep the source info around?

Yep

> I wonder whether we could change Term to be Term<T = Value>, so that when there are occasions that we can statically assert the type of the inner value, we can instead write Term<String>.

This would be great, but don't really feel like rolling that into this PR.

> Either that or we could use a different container type for values that have source info...

This would also be great; have a TODO somewhere to that effect.

Reply:
We want to do this as a follow-up.

Reply:
Task added